### PR TITLE
Add scala format.

### PR DESCRIPTION
**Pull request by sequencer**
This PR adds scalafmt (which was borrowed from firrtl).

commit 928dea8 introduced a huge code change, but it can be reapplied with
```
sbt scalafmt
```
for reviewing.

This PR depends on #197.

### CODE REVIEW

**Review by edwardcwang — COMMENTED** [review]
If you rebase against master now this PR should become easier to read.

**Comment by sequencer** [comment]
> If you rebase against master now this PR should become easier to read.

Done :)

**Comment by ducky64** [comment]
This looks fine in principle. Would like to discuss at the dev meeting about how other projects (firrtl, chisel3 - which I don't think has merged scalafmt yet, though I think it might be slated for a release version boundary) are going to handle this.

That being said, I think it will be a lot less of a problem here, since there's less open or active PRs will be impacted.

**Comment by jackkoenig** [comment]
We merged scalafmt to FIRRTL just before 1.4.0-RC1. As you mention, we decided to do it for the 1.4 major release.

**Comment by sequencer** [comment]
@ducky64, just peeked the #204, maybe this PR needs some thing to disscuss, and I think it’s not a huge one, we may don’t need to block this scalafmt PR by that? If @ducky64 think this is fine, and @ryan-lund doesn’t mind to rebase based on this PR, can we just merge this first?

**Comment by ducky64** [comment]
I'm happy with merging this whenever. Ideally would minimize pain on other PRs, but I think the only active ones are #204 and #196, which I plan to review later this week when I'm less swamped.

**Review by ducky64 — APPROVED** [review]

**Comment by sequencer** [comment]
this should be merged before #213

**Comment by sequencer** [comment]
@ducky64 @chick request to merge, just suffered from a rebase hole for #200

**Comment by ducky64** [comment]
Got it, I'd like to hold off up to another week (unless this is blocking you?) and see if we can resolve #204 and #196 first?

**Comment by ducky64** [comment]
@sequencer this is now conflicting (probably because #204 was just merged), can this get updated?

**Comment by sequencer** [comment]
Sure done:)

File: src/test/scala/chiseltest/tests/BasicTest.scala [hunk 1]
```diff
@@ -89,7 +105,7 @@ class BasicTest extends AnyFlatSpec with ChiselScalatestTester with Matchers {
       test(new StaticModule(42.U)) { c =>
         c.out.expect(0.U, "user-defined failure message =(")
       }
-    }.getMessage should include ("user-defined failure message =(")
+    }.getMessage should include("user-defined failure message =(")
```
Comment (ducky64) [inline src/test/scala/chiseltest/tests/BasicTest.scala:108]: Is there a rule that allows this to look like recommended scalatest style?

File: src/test/scala/chiseltest/tests/BundleLiteralsSpec.scala [hunk 2]
```diff
@@ -11,7 +11,7 @@ import org.scalatest.flatspec.AnyFlatSpec
 import org.scalatest.matchers.should.Matchers
 
 class BundleLiteralsSpec extends AnyFlatSpec with ChiselScalatestTester with Matchers {
-  behavior of "Testers2"
+  behavior.of("Testers2")
```
Comment (ducky64) [inline src/test/scala/chiseltest/tests/BundleLiteralsSpec.scala:14]: This is pretty bad, is there a rule to keep this as-is?

File: src/test/scala/chiseltest/tests/ClockCrossingTest.scala [hunk 3]
```diff
@@ -31,13 +31,13 @@ class ClockCrossingTest extends AnyFlatSpec with ChiselScalatestTester {
       io.mainOut := mainReg
     }) { c =>
       c.io.divIn.poke(0x42.U)
-      c.io.mainOut.expect(0.U)  // initial register value
+      c.io.mainOut.expect(0.U) // initial register value
```
Comment (ducky64) [inline src/test/scala/chiseltest/tests/ClockCrossingTest.scala:34]: Not sure I agree with one space before comment, but don't feel super strongly about it

File: src/test/scala/chiseltest/tests/FaultLocatorTest.scala [hunk 4]
```diff
@@ -39,13 +39,13 @@ class FaultLocatorTest extends AnyFlatSpec with ChiselScalatestTester with Match
...
-        c.in.bits.poke(false.B)  // Have this be a data failure only
+        c.in.bits.poke(false.B) // Have this be a data failure only
         c.out.expectDequeueNow(true.B)
       }
     }
     // Only check the filename to avoid this being too brittle as implementation changes
-    exc.failedCodeFileNameAndLineNumberString.get should startWith ("DecoupledDriver.scala:")
-    exc.getMessage should include regex ("""\(lines in FaultLocatorTest\.scala:[^\)]*43.*\)""")
+    exc.failedCodeFileNameAndLineNumberString.get should startWith("DecoupledDriver.scala:")
+    (exc.getMessage should include).regex("""\(lines in FaultLocatorTest\.scala:[^\)]*43.*\)""")
```
Comment (ducky64) [inline src/test/scala/chiseltest/tests/FaultLocatorTest.scala:48]: Yeah, the formatter did not do the right thing here either.

**Review by ducky64 — CHANGES_REQUESTED** [review]
Looked through the changes again, it looks like the formatter absolutely mangled some of the ScalaTest stuff (see inline comments). Can that be fixed?

**Comment by sequencer** [comment]
I’ll digging into scalafmt documentation for this.

**Comment by chick** [comment]
I wasn't able to find a fix for this though there may be one.
But note that firrtl and treadle have both adopted this scheme and have accepted this particular wart.

File: src/test/scala/chiseltest/tests/BundleLiteralsSpec.scala (hunk 2 above)
Comment (sequencer) [reply src/test/scala/chiseltest/tests/BundleLiteralsSpec.scala:14]: I mad a PR in scalafmt for this https://github.com/scalameta/scalafmt/pull/2252

File: src/test/scala/chiseltest/tests/ClockCrossingTest.scala (hunk 3 above)
Comment (sequencer) [reply src/test/scala/chiseltest/tests/ClockCrossingTest.scala:34]: I don’t think there are some good way to fix this.

File: src/test/scala/chiseltest/tests/FaultLocatorTest.scala (hunk 4 above)
Comment (sequencer) [reply src/test/scala/chiseltest/tests/FaultLocatorTest.scala:48]: Should I add more keyword in the PR: https://github.com/scalameta/scalafmt/pull/2252?

**Comment by sequencer** [comment]
I post a PR in scalafmt.
Here are these options:
1. Remove `AvoidInfix` to remove all these reformat(scalafmt.conf of tester2 will be inconsistent to firrtl and chisel then).
2. Append `format: off` to for some place we didn’t wanna reformat.
3. Merge this as it is, and wait for PR https://github.com/scalameta/scalafmt/pull/2252 to be merged and fix these wart in the future.

**Comment by ducky64** [comment]
Per the response in scalameta/scalafmt#2252, can you add the relevant keywords to the local exclude list in this project? I also expect that the list may grow as we have more test cases.

Alternatively, can you define special rules that apply to the test directory only? (though I don't anticipate these rules firing in non-test code?)

If this gets merged upstream later and the local rules become redundant, we can always delete them.

**Comment by sequencer** [comment]
Added 86b813e to remove `AvoidInfix` in tests.

**Comment by sequencer** [comment]
> I also expect that the list may grow as we have more test cases.

I will collect scalatests's DSL keyword and upstream them in https://github.com/scalameta/scalafmt/pull/2252, to avoid our local maintaining.

File: src/test/scala/chiseltest/tests/RegionsTest.scala [hunk 5]
```diff
@@ -18,35 +18,42 @@ class RegionsTest extends AnyFlatSpec with ChiselScalatestTester {
         c.clock.step()
         c.in.poke(70.U)
         c.clock.step()
-      }.fork.withRegion(Monitor) {
-        c.in.expect(42.U)
-        c.clock.step()
-        c.in.expect(70.U)
-      }.joinAndStep(c.clock)
+      }.fork
+        .withRegion(Monitor) {
```
Comment (ducky64) [inline src/test/scala/chiseltest/tests/RegionsTest.scala:22]: Uh, what's going on here? This looks super ugly?

File: src/test/scala/chiseltest/tests/RegionsTest.scala [hunk 6]
```diff
@@ -18,35 +18,42 @@ class RegionsTest extends AnyFlatSpec with ChiselScalatestTester {
...
-        c.in.poke(70.U)
-        c.clock.step()
-      }.joinAndStep(c.clock)
+      fork
+        .withRegion(Monitor) {
+          c.in.expect(42.U)
+          c.clock.step()
+          c.in.expect(70.U)
+        }
+        .fork {
```
Comment (ducky64) [inline src/test/scala/chiseltest/tests/RegionsTest.scala:39]: Highly undesirable... and also inconsistent with the above, where it put the fork on the close curly braces?

File: src/test/scala/chiseltest/tests/RegionsTest.scala [hunk 7]
```diff
@@ -18,35 +18,42 @@ class RegionsTest extends AnyFlatSpec with ChiselScalatestTester {
...
   }
 
   it should "not allow joining from a later region" in {
     assertThrows[TemporalParadox] {
       test(new PassthroughModule(UInt(8.W))) { c =>
-        fork.withRegion(Monitor) {
-          c.clock.step()
-        }.join()
+        fork
+          .withRegion(Monitor) {
```
Comment (ducky64) [inline src/test/scala/chiseltest/tests/RegionsTest.scala:53]: This is also pretty bad, putting withRegion on its own line

**Review by ducky64 — COMMENTED** [review]
Looks mostly fine as-is.

I found three more things (related to placement of chained `.fork`s or `.join`s) that are kind of iffy, if you want to try to fix them. Interestingly, it's limited to one file, it doesn't seem to do the wrong thing on other files?

Also, does the scalafmt PR need to go in first, or can this be merged anytime?

**Comment by ducky64** [comment]
Any progress on fixing these last issues? I think the two options discussed during the previous chisel-dev was either:
- see if it's possible to ignore formatting on fork-like constructs
- or, ignore formatting on the test directory entirely.

There's also outstanding conflicts, presumably from the latest merged PR.

**Comment by sequencer** [comment]
> see if it's possible to ignore formatting on fork-like constructs
> or, ignore formatting on the test directory entirely.

I searched and didn't find a better solution for this.
I'll ignore test directory and rebase this PR to master.

**Comment by sequencer** [comment]
In e0b8600, I remove the scalafmt for test.

**Review by ducky64 — APPROVED** [review]

**Comment by jackkoenig** [comment]
@Mergifyio backport 0.3.x

**Pull request merged**